nucliadb_telemetry/nucliadb_telemetry/logs.py

```diff
@@ -0,0 +1,192 @@
+import logging
+from copy import copy
+from datetime import datetime
+from typing import Any
+
+import orjson
+import pydantic
+from opentelemetry import trace
+from opentelemetry.trace.span import INVALID_SPAN
+
+from nucliadb_telemetry.settings import LogSettings
+
+_BUILTIN_ATTRS = (
+    # list of all possible args
+    "args",
+    "asctime",
+    "created",
+    "exc_info",
+    "exc_text",
+    "filename",
+    "funcName",
+    "levelname",
+    "levelno",
+    "lineno",
+    "module",
+    "msecs",
+    "message",
+    "msg",
+    "name",
+    "pathname",
+    "process",
+    "processName",
+    "relativeCreated",
+    "stack_info",
+    "thread",
+    "threadName",
+)
+
+_OUTPUT_PROPERTIES = {
+    "filename": "filename",
+    "funcName": "funcName",
+    "levelname": "severity",
+    "lineno": "lineno",
+    "module": "module",
+    "name": "name",
+    "pathname": "pathname",
+}
+
+
+class JSONFormatter(logging.Formatter):
+    """
+    Formatter base json-log-formatter with pydantic support
+    """
+
+    def format(self, record: logging.LogRecord) -> str:
+        extra: dict[str, Any]
+        if isinstance(record.msg, dict):
+            extra = record.msg
+        elif isinstance(record.msg, pydantic.BaseModel):
+            extra = record.msg.dict()
+        else:
+            extra = {"message": record.getMessage()}
+        extra.update(self.extra_from_record(record))
+
+        self.fill_log_data(extra, record)
+
+        return orjson.dumps(extra, default=repr).decode("utf-8", errors="ignore")
+
+    def fill_log_data(self, data: dict[str, Any], record: logging.LogRecord) -> None:
+        if "time" not in data:
+            data["time"] = datetime.utcnow()
+
+        if record.exc_info:
+            data["exc_info"] = self.formatException(record.exc_info)
+
+        for property, output_property in _OUTPUT_PROPERTIES.items():
+            data[output_property] = getattr(record, property, None)
+
+        data["logging.googleapis.com/sourceLocation"] = {
+            "file": record.pathname,
+            "line": record.lineno,
+            "function": record.funcName,
+        }
+
+        current_span = trace.get_current_span()
+        if current_span not in (INVALID_SPAN, None):
+            span_context = current_span.get_span_context()
+            # for us, this is opentelemetry trace_id/span_id
+            # GCP has logging.googleapis.com/spanId but it's for it's own cloud tracing system
+            data["trace_id"] = span_context.trace_id
+            data["span_id"] = span_context.span_id
+
+        if hasattr(record, "stack_info"):
+            data["stack_info"] = record.stack_info
+        else:  # pragma: no cover
+            data["stack_info"] = None
+
+    def extra_from_record(self, record):
+        return {
+            attr_name: record.__dict__[attr_name]
+            for attr_name in record.__dict__
+            if attr_name not in _BUILTIN_ATTRS
```

Comment:
shouldn't this be `if attr_name in _BUILTIN_ATTRS` instead?

Reply:
nah, this is to gather properties that are NOT part of the record.